### ISSUE TITLE
VerticalText epub/mobi  CSS support

### ISSUE THREAD

**Issue by kitsek** [koreader/koreader#4353]
* KOReader version: All
* Device: All

#### Issue
not supporting vertical text CSS
#### Steps to reproduce
kindle/kobo all works fine but KOREADER not working, everything vertical will become horizontal.

**Comment by Frenzie** [comment]
Incidentally, does it work if you try open with MuPDF?

This would be very hard to implement, I'm afraid.

**Comment by pazos** [comment]
Attached test epubs from https://blog.tommyku.com/blog/how-to-make-epubs-with-vertical-layout/

[horizontal.epub.zip](https://github.com/koreader/koreader/files/2605226/horizontal.epub.zip)
[vertical.epub.zip](https://github.com/koreader/koreader/files/2605227/vertical.epub.zip)

Work ok in calibre, didn't test on emulator yet.

**Comment by sekkit** [comment]
yeah, mobi/epub/kepub/azw3 with vertical css all works fine in iBooks/Kindle/Kobo etc. But KOReader's html layout engine might have issues supporting that syntax. MuPDF failed to open epub files on my MacOS10.14, for testing on Windows that will be tried later on.

**Comment by grapemix** [comment]
I got the same problem too. The epub file I tried has no problem to render text vertically by kobo's default reader, even **the epub is epub2**. But the file is not being rendered correctly by KOreader and calibre's preview viewer. Here is the css of the epub.

body {
  writing-mode: vertical-rl;
  -epub-writing-mode: vertical-rl;
  -webkit-writing-mode: vertical-rl;
  line-break: normal;
  -epub-line-break: normal;
  -webkit-line-break: normal;
}

I tired to covert the epub from epub2 to epub3 thru calibre, but the result is the same. Is KOreader support epub3?

My KOreader version is v2018.12

**Comment by Frenzie** [comment]
Not that part of EPUB3, anyway. ;-)

**Comment by poire-z** [comment]
Not sure we'll ever support that (unless we include a 3rd epub engine :)

crengine is really from the core made to handle horizontal text in the left to right direction: paragraph splitting, line layout, hyphenation, vertical stacking of rectangles and splitting to make pages, screen coordinate to HTML node (for text selection/link hitting) is scanning from left to right of these rectangles... that I don't even see how vertical text layout could fit in, without implications everywhere... (and even right-to-left text layout seem complicated).

As for the MuPDF epub engine: https://bugs.ghostscript.com/show_bug.cgi?id=695924#c16 : _We are not likely to ever support vertical writing in the EPUB viewer._

**Comment by pazos** [comment]
@poire-z: there is no 3rd epub engine. epub3 spec is a joke and readium sdk relies on a full web engine (webkit or related). I guess most commercial ebook readers (outside android/iOS) implement the render via qwebkit/qwebengine. In that case they're able to use readium **native** (:-1: ) sdk or repurpose some of the javascript present on calibre ebook viewer.

**Comment by NiLuJe** [comment]
Yep, I'm continually amazed by the fantastic mess that is the ePub format, and how fantastically badly designed it is considering its target "audience" is mainly going to be low-power devices...

**Comment by Frenzie** [comment]
While not an EPUB renderer, I think something like Prince could also do a pretty decent job. I've occasionally used it over the past decade for various purposes and I was quite satisfied with the results. You'd get rather slow first-time loading performance of course, but fast PDF rendering after.

It also supports some basic JS for when the built-in CSS properties don't quite cut it: https://fransdejonge.com/2015/08/custom-page-number-count-in-prince/

**Comment by cges30901** [comment]
> Attached test epubs from https://blog.tommyku.com/blog/how-to-make-epubs-with-vertical-layout/

I can not open the files in KOReader. KOReader version is 2019.03.1 on Linux and Android.
![螢幕快照_2019-04-03_23-03-47](https://user-images.githubusercontent.com/19516286/55489598-d1e9f580-5664-11e9-9778-0c79a0f4e53b.png)

> Work ok in calibre

No, the Chinese characters are rotated.
![螢幕快照_2019-04-03_23-05-52](https://user-images.githubusercontent.com/19516286/55489731-16759100-5665-11e9-995d-f9040c653b4a.png)

**Comment by Frenzie** [comment]
Vertical won't work. Curious about it not opening. Did you check the EPUB validator? 

http://validator.idpf.org/

**Comment by cges30901** [comment]
> Vertical won't work. Curious about it not opening. Did you check the EPUB validator?
> 
> http://validator.idpf.org/

Sorry, I can not open the link.

**Comment by cges30901** [comment]
> Vertical won't work. Curious about it not opening. Did you check the EPUB validator?
> 
> http://validator.idpf.org/

I guess there are some problems with the epub files because I can open them in KOReader after opening and saving in Sigill.

**Comment by pazos** [comment]
@cges30901: for android/ios you have [Gitden](http://gitden.com), which seems the best option for CJK and/or RTL languages.

For desktop I think some qt/webkit apps will work but I'm not sure. You can try https://pgaskin.net/ePubViewer/

**Comment by cges30901** [comment]
> Vertical won't work. Curious about it not opening. Did you check the EPUB validator?
> 
> http://validator.idpf.org/

There are some errors in the epub file.


Type | File | Line | Position | Message
-- | -- | -- | -- | --
ERROR | ./tmp/uploads/1554350319051_0000052661/inputFile/vertical.epub | - | - | The mimetype file has an  extra field of length 28. The use of the extra field feature of the ZIP  format is not permitted for the mimetype file.
ERROR | OEBPS/content.opf | 21 | 52 | Error while parsing file: attribute "page-progression-direction" not allowed here; expected attribute "id"
ERROR | OEBPS/toc.ncx | - | - | NCX identifier ('https://blog.tommyku.com/') does not match OPF identifier ('https:/blog.tommyku.com/').

**Comment by Frenzie** [comment]
Attaching the document in question for posterity.

[vertical.epub.zip](https://github.com/koreader/koreader/files/3042251/vertical.epub.zip)

I don't think any of the above should keep the document from loading. Pinging @poire-z

**Comment by poire-z** [comment]
> I don't think any of the above should keep the document from loading

Yes, it should load. It currently doesn't because of the `./` in :
`    <item id="section-one" href="./Text/Section001.xhtml" media-type="application/xhtml+xml" />`

```
ORIGINAL:    <item id="section-one" href="./Text/Section001.xhtml" media-type="application/xhtml+xml" />
2019/04/04 10:15:11.5610 INFO Authors: Tommy Ku Title: Vertical EPUB demo
2019/04/04 10:15:11.5611 DEBUG opf: reading items
2019/04/04 10:15:11.5611 DEBUG opf: reading items done.
2019/04/04 10:15:11.5611 DEBUG opf: reading spine
2019/04/04 10:15:11.5612 DEBUG opf: reading spine done
2019/04/04 10:15:11.5612 INFO ldomNode::unregisterDocument() - for index 2
2019/04/04 10:15:11.5612 DEBUG opf: closed
2019/04/04 10:15:11.5613 DEBUG setDocFlags(000f)
2019/04/04 10:15:11.5613 TRACE codeBasePrefix is empty for path
2019/04/04 10:15:11.5614 DEBUG Checking fragment: OEBPS/./Text/Section001.xhtml <= not found in the zip
2019/04/04 10:15:11.5614 TRACE codeBasePrefix is empty for path OEBPS/./
2019/04/04 10:15:11.5614 DEBUG EPUB: 0 documents merged <= so, nothing to show

When "fixed": <item id="section-one" href="Text/Section001.xhtml" media-type="application/xhtml+xml" />

2019/04/04 10:13:22.9045 INFO Authors: Tommy Ku Title: Vertical EPUB demo
2019/04/04 10:13:22.9046 DEBUG opf: reading items
2019/04/04 10:13:22.9046 DEBUG opf: reading items done.
2019/04/04 10:13:22.9046 DEBUG opf: reading spine
2019/04/04 10:13:22.9047 DEBUG opf: reading spine done
2019/04/04 10:13:22.9047 INFO ldomNode::unregisterDocument() - for index 2
2019/04/04 10:13:22.9048 DEBUG opf: closed
2019/04/04 10:13:22.9048 DEBUG setDocFlags(000f)
2019/04/04 10:13:22.9048 TRACE codeBasePrefix is empty for path
2019/04/04 10:13:22.9049 DEBUG Checking fragment: OEBPS/Text/Section001.xhtml
2019/04/04 10:13:22.9051 TRACE http-equiv content-type:  content="application/xhtml+xml; charset=utf-8" /
2019/04/04 10:13:22.9052 TRACE CSS file href: OEBPS/Style/style.css
2019/04/04 10:13:22.9052 DEBUG Setting StyleSheet attribute to OEBPS/Style/style.css for document fragment
2019/04/04 10:13:22.9053 TRACE ldomNode::applyNodeStylesheet()
2019/04/04 10:13:22.9055 TRACE codeBasePrefix is empty for path OEBPS/./
2019/04/04 10:13:22.9056 DEBUG EPUB: 1 documents merged
```

Seems fixable with:
```diff
--- a/crengine/src/epubfmt.cpp
+++ b/crengine/src/epubfmt.cpp
@@ -989,7 +989,7 @@ bool ImportEpubDocument( LVStreamRef stream, ldomDocument * m_doc, LVDocViewCall
     }
     for ( int i=0; i<spineItems.length(); i++ ) {
         if (spineItems[i]->mediaType == "application/xhtml+xml") {
-            lString16 name = codeBase + spineItems[i]->href;
+            lString16 name = LVCombinePaths(codeBase, spineItems[i]->href);
             {
                 CRLog::debug("Checking fragment: %s", LCSTR(name));
                 LVStreamRef stream = m_arc->OpenStream(name.c_str(), LVOM_READ);
--- a/crengine/src/lvstream.cpp
+++ b/crengine/src/lvstream.cpp
@@ -3783,9 +3783,11 @@ lString16 LVCombinePaths( lString16 basePath, lString16 newPath )
             }
         }
     } while ( changed && s.length()>=pattern.length() );
-    // "./"
+    // "./" at start
     if ( s.length()>2 && s[0]=='.' && s[1]==separator )
         s.erase(0, 2);
+    // "/./" => "/"
+    while ( s.replace( cs16("/./"), cs16("/") ) ) ; // or something more clever with 'separator'
     return s;
 }
```
A few more places in epubfmt.cpp could use `LVCombinePaths()` too.

After that, I don't know if it's possible that in a ZIP file, the paths stored in the content list datastructures include such `/./`, in which case this would fail again - or it needs again this replacement there when checking for matches.

Anyway, best to wait after next monthly release to introduce this "fix", in case it fails with other books were it did not before.

But strange we did not witness that before. It looks to me not so uncommon to use `./` in href...

**Comment by Frenzie** [comment]
> It looks to me not so uncommon to use ./ in href...

Maybe, maybe not. It doesn't make much sense to write it that way, in contrast to `../blabla` which has a clear use case.

Btw, shouldn't there be some standard C++ library for paths? Seems like the kind of thing you'd expect to exist.

**Comment by poire-z** [comment]
> shouldn't there be some standard C++ library for paths

There probably is. But crengine doesn't use/require any of these standard C++ libraries, and has its own home-made version of most things "standard c++" I guess (lvarray, lvhashtable, lvstring, lvref, lvqueue, lvptrvec, lvstream...).
Which I kinda like (except for these small bugs), as it allows to quickly see how this low-level stuff is implemented.

**Comment by xelxebar** [comment]
Japanese epub have given me enough grief that I'd be willing to help finding a solution to OP's issue. Have been forced to convert these to PDF/DjVu, but the inflation factor is quite large.

**Comment by Frenzie** [comment]
@xelxebar I have no idea of the complexities involved, but I would take a look at https://github.com/koreader/crengine/issues/267 for potentially some inspiration about where to start looking. It'd be a momentous undertaking in any case. But… the repo seems to have vanished?!

**Comment by poire-z** [comment]
Well, as I said above, _vertical_ text is really something different than _RTL_ horizontal text.
RTL text could somehow fit into crengine way of doing things.
But I don't see how vertical text could... (horizontal lines are added, the page splitting is really delayed, which allows scroll mode... text selection...)

**Comment by xelxebar** [comment]
Well, I'm interested in internationalization in general. Wouldn't mind working on rtl as well. I'm naive enough at this point to believe that could pave the way to finding a good vert implementation. :p

**Comment by cges30901** [comment]
Currently I am using a font that Chinese characters are rotated counterclockwise 90 degrees to workaround this issue.

Originally it is horizontal:
![0-original](https://user-images.githubusercontent.com/19516286/70152748-bdc13100-16e8-11ea-818b-fa09e183f313.png)

Using a rotated font:
![1-font](https://user-images.githubusercontent.com/19516286/70152751-bdc13100-16e8-11ea-8997-0052bb453cd0.png)

Rotate screen 90 degrees:
![2-rotate](https://user-images.githubusercontent.com/19516286/70152752-be59c780-16e8-11ea-87cb-f7d0cf69a8f1.png)

**Comment by NiLuJe** [comment]
That's one rather creative way to deal with it :D.

**Comment by pggdt** [comment]
I tried the version 2021.05 and sadly found that vertical text support is still not there. I hope one day the vertical books could be showed properly like this on Atril Document Viewer.
![1](https://user-images.githubusercontent.com/4413062/123375083-c58e4200-d5ba-11eb-8259-996caafbfdc3.png)
Thanks for your great effort!

**Comment by Sami32** [comment]
Hello everyone!

I'm still searching for the best ePUB 3.x reader on Linux but for Android i can share that "Reasily" App is by far the most compliant of them all with CSS3 and MathML support.

**Comment by NiLuJe** [comment]
I would assume the official [Readium implementation](https://github.com/edrlab/thorium-reader) would have pretty decent coverage, given who's working on it ;o).

**Comment by Sami32** [comment]
@NiLuJe Thank you very much :)

I forgot about this one because i was too focused on the poor Readium Android Apps and Calibre E-book reader standalone doesn't exist so i totally missed it.


Thanks +1

**Comment by NiLuJe** [comment]
To be fair, it's pretty damn well hidden.

**Comment by foxmean** [comment]
Not sure if it related but can we implement something like 'vertical writing mode' like [ttsu](https://github.com/ttu-ttu/ebook-reader) instead? It would be great to have options like that. You can take a look in [ttsu setting page](https://reader.ttsu.app/settings).

**Comment by poire-z** [comment]
That's related, and exactly that that is necessary and that we don't implement, and that would be really really hard to implement: https://github.com/koreader/koreader/issues/4353#issuecomment-452452412.
Also see https://github.com/koreader/koreader/pull/8312#issuecomment-974680922 and follow ups.

ttsu is just some Javascript code that deal with the book format and feed HTML/CSS to your web browser, and current web browsers can do (after years of hard work by many people) vertical text nicely. Our rendering engine (crengine) can't do that, and has not been designed to allow that easily.

And even web browsers (or ttsu) have a hard time ensuring proper rendering in vertical mode:
![ChinaVertical](https://github.com/koreader/koreader/assets/24273478/31f4b5fa-daad-47cf-8427-aa6206af08c4)
for our horizontal rendering:
![image](https://github.com/koreader/koreader/assets/24273478/ed0b2d2c-df67-4be7-8187-63aeefefd82f)

There could be some hacks, like rotating glyphs in the font - and the user rotating his device :) - and even if we were to do this glyph rotation in the code, there would be tons of other bad things like margin-top being margin-right... still a hack, not really worth integrating in the code :/

**Comment by poire-z** [comment]
Cross-ref to https://github.com/koreader/koreader/issues/11469#issuecomment-1969338188 and some user-patch that may works for others or trigger new ideas.
